### PR TITLE
Added MongoCrypt support for Azure and GCP

### PR DESCRIPTION
**Pull request by rozza**
Added calls to the new CAPI's:

  * mongocrypt_setopt_kms_providers - to support new KMS provider options.
  * mongocrypt_ctx_setopt_key_encryption_key - to support new data key options.
  * mongocrypt_setopt_crypto_hook_sign_rsaes_pkcs1_v1_5 - to support signature callback for GCP.

JAVA-3841

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java [hunk 1]
```diff
@@ -55,16 +56,21 @@ public static MongoCryptOptions createMongoCryptOptions(final Map<String, Map<St
                                 .build()
                 );
             } else {
-                throw new MongoClientException("Unrecognized KMS provider key: " + entry.getKey());
+                bsonKmsProviders.put(entry.getKey(), new BsonDocumentWrapper<>(new Document(entry.getValue()), new DocumentCodec()));
             }
         }
+
+        if (!bsonKmsProviders.isEmpty()) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java:63]: Should it be an error if there are no providers?  Or is that already detected at a level up?

File: driver-sync/src/main/com/mongodb/client/internal/Crypt.java [hunk 2]
```diff
@@ -273,7 +273,8 @@ private void mark(final MongoCryptContext cryptContext, final String databaseNam
 
     private void fetchKeys(final MongoCryptContext keyBroker) {
         try {
-            for (BsonDocument bsonDocument : keyRetriever.find(keyBroker.getMongoOperation())) {
+            RawBsonDocument mongoOperation = keyBroker.getMongoOperation();
```
Comment (jyemin) [inline driver-sync/src/main/com/mongodb/client/internal/Crypt.java:276]: This seems an unnecessary change.  Can you revert?

File: driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java [hunk 3]
```diff
@@ -85,18 +91,10 @@ public static MongoCryptOptions createMongoCryptOptions(final Map<String, Map<St
...
-                        builder.serverSelectionTimeout(1, TimeUnit.SECONDS);
-                    }
-                })
-                .applyToSocketSettings(new Block<SocketSettings.Builder>() {
-                    @Override
-                    public void apply(final SocketSettings.Builder builder) {
-                        builder.readTimeout(1, TimeUnit.SECONDS);
-                        builder.connectTimeout(1, TimeUnit.SECONDS);
-                    }
+                .applyToClusterSettings(builder -> builder.serverSelectionTimeout(1, TimeUnit.SECONDS))
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java:94]: I don't mind the occasional Java 8 syntax update, but I'd prefer them kept to separate commits so it's clear which changes are functional and which are non-functional.  Would you mind reverting and doing this separately?

File: driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java [hunk 4]
```diff
@@ -55,16 +56,21 @@ public static MongoCryptOptions createMongoCryptOptions(final Map<String, Map<St
                                 .build()
                 );
             } else {
-                throw new MongoClientException("Unrecognized KMS provider key: " + entry.getKey());
+                bsonKmsProviders.put(entry.getKey(), new BsonDocumentWrapper<>(new Document(entry.getValue()), new DocumentCodec()));
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java:59]: Clever!

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java (hunk 1 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java:63]: Handled by MongoCryptOptions

File: driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java (hunk 3 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/capi/MongoCryptHelper.java:94]: Ack

File: driver-sync/src/main/com/mongodb/client/internal/Crypt.java (hunk 2 above)
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/Crypt.java:276]: Will do

**Review by jyemin — APPROVED** [review]
LGTM!
